### PR TITLE
Smoothing for LC-MS and LC-IMS-MS

### PR DESCRIPTION
**Pull request by SteffenHeu**
- The smoothing module was completely refactored.
- added copy methods for the data series

For IMS data, i'm smoothing the summed intensities which are used to construct a regular 2d-tic. Furthermore, every mobilogram is smoothed on it's own, although the summed mobilograms are not smoothed. For now, this is intended behaviour, I'll see later on, what works during mobility deconvolution.

Feel free to leave reviews before merging. I also added/changed a bit in the ModularFeature and ModularFeatureList.

### CODE REVIEW

File: src/main/java/io/github/mzmine/datamodel/featuredata/FeatureDataUtils.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+  public static float calculateArea(IonTimeSeries<? extends Scan> series) {
+    if (series.getNumberOfValues() <= 1) {
+      return 0f;
+    }
+    float area = 0f;
+    DoubleBuffer intensities = series.getIntensityValues();
+    List<? extends Scan> scans = series.getSpectra();
+    for (int i = 1; i < series.getNumberOfValues(); i++) {
+      final double lastIntensity = intensities.get(i - 1);
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/datamodel/featuredata/FeatureDataUtils.java:121]: With this implementation each intensity is retrieved twice (once as "last" and once as "this"). Could be optimized to retrieve each value only once.

Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/featuredata/FeatureDataUtils.java:121]: You're right, good point.

**Comment by tomas-pluskal** [comment]
Looks good to me, anyone else wants to review?

File: src/main/java/io/github/mzmine/util/DataPointUtils.java [hunk 2]
```diff
@@ -44,12 +44,41 @@
...
+    for (int i = 0; i < mzValues.capacity(); i++) {
       data[0][i] = mzValues.get(i);
       data[1][i] = intensityValues.get(i);
     }
     return data;
   }
-}
\ No newline at end of file
+
+  public static double[] getDoubleBufferAsArray(DoubleBuffer values) {
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/util/DataPointUtils.java:56]: Where is this used?`We should definitely add some Java doc to layout the use cases and that people should try to use the DoubleBuffers directly - if possible

File: src/main/java/io/github/mzmine/util/DataPointUtils.java [hunk 3]
```diff
@@ -44,12 +44,41 @@
     assert mzValues.capacity() == intensityValues.capacity();
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/util/DataPointUtils.java:44]: Where is this used?`We should definitely add some Java doc to layout the use cases and that people should try to use the DoubleBuffers directly - if possible

File: src/main/java/io/github/mzmine/modules/dataprocessing/featdet_smoothing/SmoothingTask.java [hunk 4]
```diff
@@ -120,87 +118,43 @@ public void run() {
...
+              smoothedSeries = smoother.smooth(filterWeights);
+
+            } else if (featureData instanceof IonMobilogramTimeSeries) {
+              IonSpectrumSeriesSmoothing<IonMobilogramTimeSeries> smoother = new IonSpectrumSeriesSmoothing<>(
+                  ((IonMobilogramTimeSeries) featureData), newPeakList.getMemoryMapStorage(),
+                  (List<? extends MassSpectrum>) origPeakList
+                      .getSeletedScans(feature.getRawDataFile()));
+              smoothedSeries = smoother.smooth(filterWeights);
+            } else {
+              logger.info("Could not smooth feature, unknown ion series type.");
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/dataprocessing/featdet_smoothing/SmoothingTask.java:149]: maybe throw exception? or at least change to warning level logging

File: src/main/java/io/github/mzmine/modules/dataprocessing/featdet_smoothing/SmoothingTask.java [hunk 5]
```diff
@@ -111,7 +109,7 @@ public void run() {
       final double[] filterWeights = SavitzkyGolayFilter.getNormalizedWeights(filterWidth);
 
       // Create new feature list
-      newPeakList = new ModularFeatureList(origPeakList + " " + suffix, origPeakList.getRawDataFiles());
+      newPeakList = origPeakList.createEmptyCopyWithSameTypes(origPeakList + " " + suffix);
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/dataprocessing/featdet_smoothing/SmoothingTask.java:112]: I think it is better to copy the whole feature list and change the feature data in this task.
Further down, you use the new ModularFeature(oldFeature, flist) constructor which copies the feature anyways - however, you do not take care of FeatureListRows. I supsect that the manual annotations and other RowTypes are not copied right now.

The featurelsit.copy method is fast and we can make sure that we do not miss anything.

File: src/main/java/io/github/mzmine/modules/dataprocessing/featdet_smoothing/SmoothingTask.java [hunk 6]
```diff
@@ -120,87 +118,43 @@ public void run() {
 
           // Create a new peak-list row.
           final int originalID = row.getID();
-          final FeatureListRow newRow = new ModularFeatureListRow(newPeakList, originalID);
+          final ModularFeatureListRow newRow = new ModularFeatureListRow(newPeakList, originalID);
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/dataprocessing/featdet_smoothing/SmoothingTask.java:121]: See earlier comment. newRow does not inherit all values from  the old row

File: src/main/java/io/github/mzmine/modules/dataprocessing/featdet_smoothing/SmoothingTask.java [hunk 7]
```diff
@@ -120,87 +118,43 @@ public void run() {
...
+                      .getSeletedScans(feature.getRawDataFile()));
+              smoothedSeries = smoother.smooth(filterWeights);
+            } else {
+              logger.info("Could not smooth feature, unknown ion series type.");
+              continue;
             }
+
+            ModularFeature newFeature = new ModularFeature(newPeakList, feature);
+            newFeature.set(FeatureDataType.class, smoothedSeries);
+            FeatureDataUtils.recalculateIonSeriesDependingTypes(newFeature);
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/dataprocessing/featdet_smoothing/SmoothingTask.java:155]: See earlier comment. newFeature is a copy of the old feature - this is pretty much what the featureList.copy method does

**Comment by SteffenHeu** [comment]
@robinschmid should take a short look at the feature list and the feature, I think.

File: src/main/java/io/github/mzmine/datamodel/features/ModularFeatureList.java [hunk 8]
```diff
@@ -461,31 +489,49 @@ public void setDateCreated(String date) {
...
+    for (FeatureListAppliedMethod proc : this.getAppliedMethods()) {
+      flist.addDescriptionOfAppliedTask(proc);
+    }
+
+    selectedScans.forEach(flist::setSelectedScans);
+    return flist;
+  }
+
+  public ModularFeatureList createEmptyCopyWithSameTypes(String title) {
+    ModularFeatureList flist = new ModularFeatureList(title, getRawDataFiles());
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/datamodel/features/ModularFeatureList.java:512]: See comment in SmoothingTask. I think we do not need such functionality. The risk of loosing any information is high.

File: src/main/java/io/github/mzmine/datamodel/features/ModularFeature.java [hunk 9]
```diff
@@ -272,7 +317,7 @@ public DataPoint getDataPoint(Scan scan) {
     if (index < 0) {
       return null;
     }
-    return getDataPoints().get(index);
+    return new SimpleDataPoint(getFeatureData().getMZ(index), getFeatureData().getIntensity(index));
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/datamodel/features/ModularFeature.java:320]: We should at least deprecate this method or try to remove it completely

File: src/main/java/io/github/mzmine/util/DataPointUtils.java (hunk 3 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/util/DataPointUtils.java:44]: I just never know when it will show multiple lines and when it only shows one :D

Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/util/DataPointUtils.java:44]: When copying the series which store the data. The MemoryMapStorage stores arrays, so we have to load them first and then store the new array.

File: src/main/java/io/github/mzmine/util/DataPointUtils.java (hunk 2 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/util/DataPointUtils.java:56]: During smoothing you keep the old mzs, but store new intensities. So loading just a single buffer is useful,too.

File: src/main/java/io/github/mzmine/util/DataPointUtils.java (hunk 3 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/util/DataPointUtils.java:44]: > 
> 
> Where is this used?`We should definitely add some Java doc to layout the use cases and that people should try to use the DoubleBuffers directly - if possible

True. However, I don't exactly know, how the memory mapped file works and if it's faster to query all points at once at some point.

File: src/main/java/io/github/mzmine/datamodel/features/ModularFeatureList.java (hunk 8 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/features/ModularFeatureList.java:512]: Will remove again.

File: src/main/java/io/github/mzmine/datamodel/features/ModularFeature.java (hunk 9 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/features/ModularFeature.java:320]: The problem is that the Feature interface still has it. As soon was we remove that, we can remove this method. I went through all usages from ModularFeature in #154.

**Comment by tomas-pluskal** [comment]
Thanks Steffen!

**Pull request merged**